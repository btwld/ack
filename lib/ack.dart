library;

export 'src/ack.dart';
