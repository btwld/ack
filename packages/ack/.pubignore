# Development and IDE files
melos_ack.iml
.claude/

# Build artifacts
build/
coverage/

# Local example (if any development examples exist)
example/
