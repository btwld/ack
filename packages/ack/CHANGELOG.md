## 0.2.0-beta.1 (2025-05-03)

* See [release notes](https://github.com/btwld/ack/releases/tag/v0.2.0-beta.1) for details.


## 0.2.0-beta.1

 - Bump "ack" to `0.2.0-beta.1`.

## 0.2.0

 - Bump "ack" to 0.2.0 with improved SchemaModel API and enhanced string validation

## 0.2.0

### Breaking Changes

- **SchemaModel API**: Completely redesigned the SchemaModel API for a more intuitive and streamlined experience
  - Changed constructor parameter from `Map<String, dynamic>` to `Object?` for better flexibility
  - Added automatic validation during construction
  - Added `isValid` getter and `getErrors()` method for easier error handling
  - Added abstract `getSchema()` method to define the schema
  - Removed `parse()`, `fromValidated()`, `validateMap()`, `validate()`, `createFromMap()`, and `initialize()` methods

### Improvements

- Improved error handling with more detailed error messages
- Better type safety with `Object?` instead of `dynamic`
- Simplified API with fewer methods and more intuitive usage
- **String Validation**: Enhanced string validation with improved methods
  - Changed `matches()` to validate full string patterns (anchored with ^ and $)
  - Added `contains()` method for partial string matching

## 0.1.2

- Improved error messages

## 0.1.1

-  Added deprecations

## 0.1.0

-  Error and exception improvements
-  Reworked validation and constraint workflows
-  Improved testing

## 0.0.2

-  Better JSON response parsing for OpenApiSchemaConverter

## 0.0.1

-  Initial version.