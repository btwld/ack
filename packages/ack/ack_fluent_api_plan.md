# Ack Fluent API Refactor Plan

This document outlines the plan to implement a fluent, Zod-inspired API for the Ack validation library. The work is broken down by schema type and is based on a detailed analysis of our current architecture and the capabilities of the Zod library.

---

### **Guiding Principles**

1.  **Universal Methods go on the Base Class**: Any capability that can apply to *any* schema (like making it nullable or adding a description) should be a method on the base `AckSchema`.
2.  **Specific Methods go in Extensions**: Capabilities that only make sense for a specific type (e.g., email validation for strings) should be implemented as extension methods on that specific schema class (`StringSchema`, `ListSchema`, etc.). This keeps the core classes clean.
3.  **Unsupported Features**: We will clearly mark features from Zod that we cannot or should not implement right now due to significant architectural differences (e.g., transformations, async validation).

---

### **1. `AckSchema` (Base Class Methods)**

These methods will be added directly to the `AckSchema<T>` class because they are universal.

| Method Signature | Zod Equivalent | Why Here? | Implementation Notes |
| :--- | :--- | :--- | :--- |
| `nullable()` | `.nullable()` | Makes any schema type nullable. This is the most fundamental and universal operation. | Already implemented via the `.nullable()` extension. No new work needed. |
| `default(T value)` | `.default(value)` | Sets a default value if the input is `null` or `undefined`. A universal concept. | Already implemented via the `defaultValue` constructor parameter. The `copyWith` pattern allows us to set it fluently. |
| `description(String text)` | `.describe(text)` | Adds a description for documentation/error message purposes. Universal. | Already implemented via the `description` constructor parameter. |

**Conclusion:** The most fundamental universal methods are already supported by our new architecture.

---

### **2. `StringSchema` Extensions**

These will be implemented in a new file: `lib/src/schemas/extensions/string_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `minLength(int n)` | `.min(n)` | String-specific length validation. | **Can Support Now** |
| `maxLength(int n)` | `.max(n)` | String-specific length validation. | **Can Support Now** |
| `length(int n)` | `.length(n)` | String-specific length validation. | **Can Support Now** |
| `email()` | `.email()` | String-specific format validation. | **Can Support Now** |
| `url()` | `.url()` | String-specific format validation. | **Can Support Now** |
| `uuid()` | `.uuid()` | String-specific format validation. | **Can Support Now** |
| `cuid()` | `.cuid()` | String-specific format validation. | **Can Support Now** (less common, maybe lower priority) |
| `datetime()` | `.datetime()` | String-specific format validation. | **Can Support Now** |
| `regex(RegExp re)` | `.regex(re)` | String-specific pattern matching. | **Can Support Now** |
| `startsWith(String s)` | `.startsWith(s)` | String-specific content validation. | **Can Support Now** |
| `endsWith(String s)` | `.endsWith(s)` | String-specific content validation. | **Can Support Now** |
| `trim()` | `.trim()` | **Transformation**. | **Cannot Support Now**. Our architecture validates, it doesn't transform the input value. |
| `toLowerCase()` | `.toLowerCase()` | **Transformation**. | **Cannot Support Now**. |
| `toUpperCase()` | `.toUpperCase()` | **Transformation**. | **Cannot Support Now**. |

---

### **3. `IntegerSchema` & `DoubleSchema` Extensions**

These will be in `lib/src/schemas/extensions/num_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `gt(num n)` / `greaterThan(num n)` | `.gt(n)` | Numeric comparison. | **Can Support Now** |
| `gte(num n)` / `min(num n)` | `.gte(n)` | Numeric comparison. | **Can Support Now** |
| `lt(num n)` / `lessThan(num n)` | `.lt(n)` | Numeric comparison. | **Can Support Now** |
| `lte(num n)` / `max(num n)` | `.lte(n)` | Numeric comparison. | **Can Support Now** |
| `isPositive()` | `.positive()` | Numeric sign check. | **Can Support Now** |
| `isNegative()` | `.negative()` | Numeric sign check. | **Can Support Now** |
| `isNonPositive()` | `.nonpositive()` | Numeric sign check. | **Can Support Now** |
| `isNonNegative()` | `.nonnegative()` | Numeric sign check. | **Can Support Now** |
| `multipleOf(num n)` / `step(num n)` | `.multipleOf(n)` | Numeric divisibility check. | **Can Support Now** |
| `finite()` | `.finite()` | Checks if a number is finite. | **Can Support Now** (More relevant for `DoubleSchema`) |
| `safe()` | `.safe()` | Checks if an integer is a "safe" integer. | **Can Support Now** (For `IntegerSchema`) |

---

### **4. `ListSchema` Extensions**

These will be in `lib/src/schemas/extensions/list_schema_extensions.dart`.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `minLength(int n)` | `.min(n)` | List-specific length validation. | **Can Support Now** |
| `maxLength(int n)` | `.max(n)` | List-specific length validation. | **Can Support Now** |
| `length(int n)` | `.length(n)` | List-specific length validation. | **Can Support Now** |
| `nonempty()` | `.nonempty()` | A convenient shorthand for `.minLength(1)`. | **Can Support Now** |

---

### **5. `ObjectSchema` Extensions**

This is where things get more complex. Our `ObjectSchema` is powerful but doesn't have all of Zod's dynamic capabilities.

| Method Signature | Zod Equivalent | Why Here? | Support Level |
| :--- | :--- | :--- | :--- |
| `pick(List<String> keys)` | `.pick()` | **Structural Transformation**. | **Cannot Support Now**. This creates a *new* schema type by altering the properties. Our `copyWith` can't do this. |
| `omit(List<String> keys)` | `.omit()` | **Structural Transformation**. | **Cannot Support Now**. Same reason as `pick`. |
| `partial()` | `.partial()` | **Structural Transformation**. | **Cannot Support Now**. This changes the `required` status of all properties, which would require deep schema cloning. |
| `deepPartial()` | `.deepPartial()` | **Structural Transformation**. | **Cannot Support Now**. Even more complex than `partial`. |
| `merge(ObjectSchema other)` | `.merge()` | Combines two object schemas. | **Can Support Now (with limitations)**. We can implement a version of this that merges properties and `required` fields. |
| `passthrough()` | `.`passthrough()` | Allows unknown keys. | **Already Supported**. This is our `allowAdditionalProperties: true` flag. We could add a fluent method for it. |
| `strict()` | `.strict()` | Disallows unknown keys. | **Already Supported**. This is our `allowAdditionalProperties: false` flag. We could add a fluent method for it. |
| `catchall(AckSchema valueSchema)` | `.catchall()` | Validates all unknown keys against a schema. | **Cannot Support Now**. Our current implementation only checks for the *presence* of additional properties, not their *values*. |

---

### **6. Features We Cannot Support (The "Why")**

*   **Transformations (`.trim()`, `.toLowerCase()`, `.transform()`, `.refine()`)**:
    *   **Reason:** Our architecture has a clear separation of concerns: a schema's job is to **validate** an input and return a `SchemaResult` with either the original (but correctly typed) value or an error. It does not **mutate or transform** the input value. Zod's `.transform()` and `.refine()` fundamentally change the output value, which is a different paradigm. Supporting this would require a major redesign of the `parseAndValidate` flow.

*   **Structural Transformations (`.pick()`, `.omit()`, `.partial()`)**:
    *   **Reason:** These methods in Zod dynamically create *new* schema definitions at runtime by introspecting and cloning existing schemas. Our `copyWith` pattern is great for creating a modified *copy*, but it's not designed for this level of dynamic, structural alteration (e.g., removing keys from the `properties` map and the `requiredProperties` list simultaneously). It's possible, but would require significantly more complex logic in the `ObjectSchema`'s `copyWith` method.

*   **Asynchronous Validation (`.refineAsync()`)**:
    *   **Reason:** Our entire validation pipeline (`validate` -> `parseAndValidate` -> `SchemaResult`) is synchronous. Introducing `async` validation would require making the entire pipeline return `Future<SchemaResult>`, which is a massive and breaking change across the entire library. 