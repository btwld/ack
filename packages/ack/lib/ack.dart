/// ACK - A Comprehensive Kotlin-inspired validation library for Dart
library;

// Constraints
export 'src/constraints/constraint.dart';
// Context
export 'src/context.dart';
export 'src/schemas/extensions/list_schema_extensions.dart';
export 'src/schemas/extensions/numeric_extensions.dart';
export 'src/schemas/extensions/object_schema_extensions.dart';
// Extensions
export 'src/schemas/extensions/string_schema_extensions.dart';
// Core schemas
export 'src/schemas/schema.dart';
export 'src/validation/ack_exception.dart';
export 'src/validation/schema_error.dart';
// Validation results
export 'src/validation/schema_result.dart';
