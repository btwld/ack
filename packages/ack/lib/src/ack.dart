import 'schemas/extensions/string_schema_extensions.dart';
import 'schemas/schema.dart';

/// The main entry point for creating schemas with the Ack validation library.
///
/// Provides a fluent API for creating various schema types.
final class Ack {
  /// Creates a string schema.
  static StringSchema string() => const StringSchema();

  /// Creates a literal string schema that only accepts the exact [value].
  /// Similar to Zod's `z.literal("value")`.
  static StringSchema literal(String value) => string().literal(value);

  /// Creates an integer schema.
  static IntegerSchema integer() => const IntegerSchema();

  /// Creates a double schema.
  static DoubleSchema double() => const DoubleSchema();

  /// Creates a boolean schema.
  static BooleanSchema boolean() => const BooleanSchema();

  /// Creates an object schema with the given properties.
  static ObjectSchema object(
    Map<String, AckSchema> properties, {
    List<String> required = const [],
    bool additionalProperties = false,
  }) =>
      ObjectSchema(
        properties.cast(),
        required: required,
        additionalProperties: additionalProperties,
      );

  /// Creates a discriminated object schema for polymorphic validation.
  static DiscriminatedObjectSchema discriminated({
    required String discriminatorKey,
    required Map<String, ObjectSchema> schemas,
  }) =>
      DiscriminatedObjectSchema(
        discriminatorKey: discriminatorKey,
        schemas: schemas,
      );

  /// Creates a list schema with the given item schema.
  static ListSchema<T> list<T extends Object>(AckSchema<T> itemSchema) =>
      ListSchema(itemSchema);

  /// Creates an enum schema for validating enum values.
  static EnumSchema<T> enumValues<T extends Enum>(List<T> values) =>
      EnumSchema(values: values);

  /// Creates a schema that can be one of many types.
  static AnyOfSchema anyOf(List<AckSchema> schemas) => AnyOfSchema(schemas);
}
