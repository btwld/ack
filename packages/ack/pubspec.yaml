name: ack
description: A simple validation library for Dart
version: 0.2.0
repository: https://github.com/btwld/ack

environment:
  sdk: ^3.3.0

dependencies:
  meta: ^1.15.0

dev_dependencies:
  coverage: ^1.11.1
  dart_code_metrics_presets: ^2.22.0
  lints: ^5.0.0
  test: ^1.24.0
