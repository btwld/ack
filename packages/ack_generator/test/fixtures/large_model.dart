import 'package:ack/ack.dart';

part 'large_model.g.dart';

/// Large model with many properties for performance testing
@Schema(
  description: 'Large model with many properties for performance testing',
  additionalProperties: true,
  additionalPropertiesField: 'extraData',
)
class LargeModel {
  // String fields with various validations
  @IsNotEmpty()
  final String field1;
  
  @IsEmail()
  final String field2;
  
  @MinLength(5)
  final String field3;
  
  @IsNotEmpty()
  final String field4;
  
  @MinLength(3)
  final String field5;
  
  // Numeric fields
  final int field6;
  final double field7;
  final int field8;
  final double field9;
  final int field10;
  
  // Boolean fields
  final bool field11;
  final bool field12;
  final bool field13;
  
  // DateTime fields
  final DateTime field14;
  final DateTime field15;
  
  // Optional fields with various constraints
  @Nullable()
  final String? field16;
  
  @Required()
  final String? field17;
  
  @Nullable()
  final int? field18;
  
  @Required()
  final bool? field19;
  
  @Nullable()
  final DateTime? field20;
  
  // Collection fields
  final List<String> field21;
  final List<int> field22;
  final Map<String, dynamic> field23;
  final List<bool> field24;
  
  // Additional properties field
  final Map<String, dynamic> extraData;

  LargeModel({
    required this.field1,
    required this.field2,
    required this.field3,
    required this.field4,
    required this.field5,
    required this.field6,
    required this.field7,
    required this.field8,
    required this.field9,
    required this.field10,
    required this.field11,
    required this.field12,
    required this.field13,
    required this.field14,
    required this.field15,
    this.field16,
    this.field17,
    this.field18,
    this.field19,
    this.field20,
    required this.field21,
    required this.field22,
    required this.field23,
    required this.field24,
    Map<String, dynamic>? extraData,
  }) : extraData = extraData ?? {};
}
