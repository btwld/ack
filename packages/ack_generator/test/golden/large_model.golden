// GENERATED CODE - DO NOT MODIFY BY HAND

// ignore_for_file: unnecessary_this, unnecessary_new, unnecessary_const, prefer_collection_literals
// ignore_for_file: lines_longer_than_80_chars, unnecessary_null_checks, non_constant_identifier_names

part of 'large_model.dart';

// **************************************************************************
// AckSchemaGenerator
// **************************************************************************

/// Generated schema for LargeModel
/// Large model with many properties for performance testing
class LargeModelSchema extends BaseSchema<LargeModelSchema> {
  /// Default constructor for parser instances
  const LargeModelSchema();

  /// Private constructor for validated instances
  const LargeModelSchema._valid(Map<String, Object?> data) : super.valid(data);

  static final ObjectSchema schema = Ack.object(
    {
      'field1': Ack.string.notEmpty(),
      'field2': Ack.string.email(),
      'field3': Ack.string.minLength(5),
      'field4': Ack.string.notEmpty(),
      'field5': Ack.string.minLength(3),
      'field6': Ack.int,
      'field7': Ack.double,
      'field8': Ack.int,
      'field9': Ack.double,
      'field10': Ack.int,
      'field11': Ack.boolean,
      'field12': Ack.boolean,
      'field13': Ack.boolean,
      'field14': DateTimeSchema.schema,
      'field15': DateTimeSchema.schema,
      'field16': Ack.string.nullable(),
      'field17': Ack.string.nullable(),
      'field18': Ack.int.nullable(),
      'field19': Ack.boolean.nullable(),
      'field20': DateTimeSchema.schema.nullable(),
      'field21': Ack.list(Ack.string),
      'field22': Ack.list(Ack.int),
      'field23': Ack.object({}, additionalProperties: true),
      'field24': Ack.list(Ack.boolean),
    },
    required: [
      'field1',
      'field2',
      'field3',
      'field4',
      'field5',
      'field6',
      'field7',
      'field8',
      'field9',
      'field10',
      'field11',
      'field12',
      'field13',
      'field14',
      'field15',
      'field17',
      'field19',
      'field21',
      'field22',
      'field23',
      'field24'
    ],
    additionalProperties: true,
  );

  /// Parse with validation - core implementation
  @override
  LargeModelSchema parse(Object? data) {
    final result = definition.validate(data);
    if (result.isOk) {
      final validatedData = Map<String, Object?>.from(
        result.getOrThrow(),
      );
      return LargeModelSchema._valid(validatedData);
    }
    throw AckException(result.getError());
  }

  /// Ensures this schema and its dependencies are registered
  static void ensureInitialize() {
    SchemaRegistry.register<LargeModelSchema>(
      (data) => const LargeModelSchema().parse(data),
    );
    // Register schema dependencies
    DateTimeSchema.ensureInitialize();
  }

  @override
  ObjectSchema get definition => schema;

  String get field1 => getValue<String>('field1')!;

  String get field2 => getValue<String>('field2')!;

  String get field3 => getValue<String>('field3')!;

  String get field4 => getValue<String>('field4')!;

  String get field5 => getValue<String>('field5')!;

  int get field6 => getValue<int>('field6')!;

  double get field7 => getValue<double>('field7')!;

  int get field8 => getValue<int>('field8')!;

  double get field9 => getValue<double>('field9')!;

  int get field10 => getValue<int>('field10')!;

  bool get field11 => getValue<bool>('field11')!;

  bool get field12 => getValue<bool>('field12')!;

  bool get field13 => getValue<bool>('field13')!;

  DateTimeSchema get field14 {
    return const DateTimeSchema()
        .parse(getValue<Map<String, Object?>>('field14')!);
  }

  DateTimeSchema get field15 {
    return const DateTimeSchema()
        .parse(getValue<Map<String, Object?>>('field15')!);
  }

  String? get field16 => getValue<String>('field16');

  String? get field17 => getValue<String>('field17');

  int? get field18 => getValue<int>('field18');

  bool? get field19 => getValue<bool>('field19');

  DateTimeSchema? get field20 {
    final data = getValue<Map<String, Object?>>('field20');
    return data != null ? const DateTimeSchema().parse(data) : null;
  }

  List<String> get field21 {
    return getValue<List>('field21')!.cast<String>();
  }

  List<int> get field22 {
    return getValue<List>('field22')!.cast<int>();
  }

  Map<String, dynamic> get field23 =>
      getValue<Map<String, Object?>>('field23')!;

  List<bool> get field24 {
    return getValue<List>('field24')!.cast<bool>();
  }

  Map<String, Object?> get extraData {
    final map = toMap();
    final knownFields = {
      'field1',
      'field2',
      'field3',
      'field4',
      'field5',
      'field6',
      'field7',
      'field8',
      'field9',
      'field10',
      'field11',
      'field12',
      'field13',
      'field14',
      'field15',
      'field16',
      'field17',
      'field18',
      'field19',
      'field20',
      'field21',
      'field22',
      'field23',
      'field24'
    };
    return Map.fromEntries(
        map.entries.where((e) => !knownFields.contains(e.key)));
  }

  /// Convert the schema to a JSON Schema
  static Map<String, Object?> toJsonSchema() =>
      JsonSchemaConverter(schema: schema).toSchema();
}
